Docker improvements.

jsha opened the pull request:
Use bridged networking.

Add some files to .dockerignore to shrink the build state sent to Docker
daemon.

Use specific hostnames to contact services, rather than localhost.

Add instructions for adding those hostnames to /etc/hosts in non-Docker config.

Use DSN-style connect strings for DBs.

Remove localhost / 127.0.0.1 rewrite hack from create_db.sh.

Add hosts section with new hostnames.

Remove bin from .dockerignore.

SQL grants go to %

Short-circuit DB creation if already existing.

Make `go install` a part of Docker image build so that Docker run is much
faster.

Bind to 0.0.0.0 for OCSP responders so they can be reached from host, and
publish / expose their ports.

Remove ToSServerThread and test.js' fetch of ToS.

Increase the registrationsPerIP rate limit threshold. When issuing from a Docker
host, the 127.0.0.1 override doesn't apply, so the limit is quickly hit.

Update docker-compose for bridged networking. Note: docker-compose doesn't currently work, but should be close.

Alternative to #1621. cc @hlandau @ibukanov @jfrazelle

ibukanov commented on test/run-docker.sh:
```diff
@@ -34,18 +34,15 @@ if [[ "$(is_running boulder-mysql)" != "true" ]]; then
 	# 3306 with host networking
 	docker run -d \
-		--net host \
+		-p 3306 \
```
As mariadb is only used for internal containers, the port should not be exposed. Otherwise mariadb will be reachable again from the outside world with passwordless setup. Similarly with boulder-rabbitmq.

ibukanov commented on test/sa_db_users.sql:
```diff
@@ -15,48 +15,48 @@
-GRANT SELECT,INSERT,UPDATE ON registrations TO 'sa'@'localhost';
-GRANT SELECT,INSERT,UPDATE ON challenges TO 'sa'@'localhost';
-GRANT SELECT,INSERT on fqdnSets TO 'sa'@'localhost';
+GRANT SELECT,INSERT,UPDATE ON authz TO 'sa'@'boulder';
```
Do you have by chance docker 1.8 that due to a bug exposed all the containers in /etc/hosts? Otherwise I do not see how this could work. The name boulder should not be known to mariadb container so mysql could not perform host name lookup. I suggest to simply replace the host with * when mysql runs inside the container for boulder.

jfrazelle commented:
If you want to do this without having to change all the host names and you
are running I believe docker > 1.9 you can run them all in the same net
namespace then 127.0.0.1 will resolve in the container properly

It's --net container:first_container_name
To reuse the net namespace for another container

On Tuesday, March 22, 2016, Coveralls <notifications@github.com> wrote:

> [image: Coverage Status] <https://coveralls.io/builds/5504641>
>
> Coverage decreased (-21.0%) to 31.3% when pulling *35770a0
> <https://github.com/letsencrypt/boulder/commit/35770a057cac9213121eef27ec888a76d40ae3db>
> on docker-fixup* into *c7ba6f1
> <https://github.com/letsencrypt/boulder/commit/c7ba6f13d2d36e64ab13f2a3ea987127eda6a4df>
> on master*.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199843850>
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

jsha commented:
@jfrazelle: tell me more about how to put everything in the same net namespace? Or is everything is the same net namespace by default?

ibukanov commented:
One can reuse the network stack of another container since docker 1.0
if not earlier. When doing so it is much better to have a separated
container running sleep inf or similar to avoid loosing network on
container restart.

On 22/03/2016, Jess Frazelle <notifications@github.com> wrote:
> If you want to do this without having to change all the host names and you
> are running I believe docker > 1.9 you can run them all in the same net
> namespace then 127.0.0.1 will resolve in the container properly
>
> It's --net container:first_container_name
> To reuse the net namespace for another container
>
> On Tuesday, March 22, 2016, Coveralls <notifications@github.com> wrote:
>
>> [image: Coverage Status] <https://coveralls.io/builds/5504641>
>>
>> Coverage decreased (-21.0%) to 31.3% when pulling *35770a0
>> <https://github.com/letsencrypt/boulder/commit/35770a057cac9213121eef27ec888a76d40ae3db>
>> on docker-fixup* into *c7ba6f1
>> <https://github.com/letsencrypt/boulder/commit/c7ba6f13d2d36e64ab13f2a3ea987127eda6a4df>
>> on master*.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199843850>
>>
>
>
> --
>
>
> Jessie Frazelle
> 4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
> pgp.mit.edu
> <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
>
>
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199850191

jfrazelle commented:
Yeah no prob sorry will send a diff for you to try (was responding from my
phone before ;)

On Tuesday, March 22, 2016, Igor Bukanov <notifications@github.com> wrote:

> One can reuse the network stack of another container since docker 1.0
> if not earlier. When doing so it is much better to have a separated
> container running sleep inf or similar to avoid loosing network on
> container restart.
>
> On 22/03/2016, Jess Frazelle <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> > If you want to do this without having to change all the host names and
> you
> > are running I believe docker > 1.9 you can run them all in the same net
> > namespace then 127.0.0.1 will resolve in the container properly
> >
> > It's --net container:first_container_name
> > To reuse the net namespace for another container
> >
> > On Tuesday, March 22, 2016, Coveralls <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> >
> >> [image: Coverage Status] <https://coveralls.io/builds/5504641>
> >>
> >> Coverage decreased (-21.0%) to 31.3% when pulling *35770a0
> >> <
> https://github.com/letsencrypt/boulder/commit/35770a057cac9213121eef27ec888a76d40ae3db
> >
> >> on docker-fixup* into *c7ba6f1
> >> <
> https://github.com/letsencrypt/boulder/commit/c7ba6f13d2d36e64ab13f2a3ea987127eda6a4df
> >
> >> on master*.
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly or view it on GitHub
> >> <
> https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199843850>
> >>
> >
> >
> > --
> >
> >
> > Jessie Frazelle
> > 4096R / D4C4 DD60 0D66 F65A 8EFC 511E 18F3 685C 0022 BFF3
> > pgp.mit.edu
> > <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
> >
> >
> > ---
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub:
> > https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199850191
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/letsencrypt/boulder/pull/1639#issuecomment-199888772>
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

jfrazelle commented:
https://github.com/letsencrypt/boulder/compare/master...jfrazelle:use-same-namespace?expand=1

jsha commented:
I decided to go ahead and keep the name changes (i.e. not use net namespacing), since this reflects more closely what we do in prod - different hostnames for different services.

I removed the `-p` flags and fixed a few things related to unittests on Travis, so tests should pass now.

@ibukanov @jfrazelle @hlandau, take a second look?

jfrazelle commented:
will try it out!

On Sat, Mar 26, 2016 at 1:29 PM, Jacob Hoffman-Andrews <
notifications@github.com> wrote:

> I decided to go ahead and keep the name changes (i.e. not use net
> namespacing), since this reflects more closely what we do in prod -
> different hostnames for different services.
>
> I removed the -p flags and fixed a few things related to unittests on
> Travis, so tests should pass now.
>
> @ibukanov <https://github.com/ibukanov> @jfrazelle
> <https://github.com/jfrazelle> @hlandau <https://github.com/hlandau>,
> take a second look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/letsencrypt/boulder/pull/1639#issuecomment-201926423>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

ibukanov commented on test/create_db.sh:
```diff
@@ -37,12 +42,11 @@ for svc in $SERVICES; do
-				sed -e "s/'localhost'/'127.0.0.1'/g" < $USERS_SQL | \
-					mysql $dbconn -D $db || \
-					die "unable to add users to ${db}"
+				sed -e "s/'localhost'/'172.%'/g" < $USERS_SQL | \
```
Any reasons to use 172.% pattern rather than just %? Given that the mysql container runs with MYSQL_ALLOW_EMPTY_PASSWORD it cannot be used as a way to provide any real security benefits like preventing accidental exposure to the internet of a container running with ipv6 enabled and a routable ipv6 address. On the other hand 172.% assumes too much about docker networking details and may need to change on future docker updates.

ibukanov commented on test/run-docker.sh:
```diff
@@ -34,18 +34,13 @@ if [[ "$(is_running boulder-mysql)" != "true" ]]; then
 		-e MYSQL_ALLOW_EMPTY_PASSWORD=yes \
 		--name boulder-mysql \
-		mariadb:10.0 mysqld --bind-address=127.0.0.1
+		mariadb:10.0 mysqld
```
The patch should use here --bind-address=0.0.0.0. Otherwise mysqld listens on all IPv4 and IPv6 addresses so when docker runs with IPv6 enabled that assigns globally routable IPv6 addresses to the container the container will be exposed to the Internet again. This is due to unfortunate feature with Docker ipv6 implementation that does not block connections to unpublished ports from outside world.

ibukanov commented on test/run-docker.sh:
```diff
@@ -34,18 +34,13 @@ if [[ "$(is_running boulder-mysql)" != "true" ]]; then
-	# with host networking
 	docker run -d \
-		--net host \
-		-e RABBITMQ_NODE_IP_ADDRESS=127.0.0.1 \
```
For the same reasons as with mariadb image the patch should not expose rabbitmq over IPv6. If I understand https://www.rabbitmq.com/networking.html correctly for that it is sufficient to specify `RABBITMQ_NODE_IP_ADDRESS=0.0.0.0` here.

jsha commented:
@ibukanov: Good points, all fixed. Thanks!

jsha commented:
@ibukanov @hlandau @jfrazelle: Any more comments on this? I'd like to merge it soon-ish and then do another rev to move to MariaDB 10.1.

One currently extant problem: This no longer works with docker-compose due to the change in hostnames. Docker-compose seems to not like dashes in hostnames, but when I tried changing them to not have dashes, the hosts still couldn't communicate with each other.

ibukanov commented:
I do not see any issue with the patch. As for docker-compose, was it updated it to include the links, https://docs.docker.com/compose/compose-file/#links ?

jsha commented:
Thanks!

> As for docker-compose, was it updated it to include the links, https://docs.docker.com/compose/compose-file/#links ?

Yep. I'll go ahead and push up the work I had in docker-compose, then get a final review from Roland or Kane and merge.

jfrazelle commented:
It looks good to me!

On Mon, Apr 4, 2016 at 2:19 PM, Jacob Hoffman-Andrews <
notifications@github.com> wrote:

> Thanks!
>
> As for docker-compose, was it updated it to include the links,
> https://docs.docker.com/compose/compose-file/#links ?
>
> Yep. I'll go ahead and push up the work I had in docker-compose, then get
> a final review from Roland or Kane and merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/letsencrypt/boulder/pull/1639#issuecomment-205501833>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

jsha commented:
@rolandshoemaker @riking could one of you review now? Thanks!

ibukanov commented:
With this change it is no longer possible to run a standalone letsencrypt client that runs on the host and binds to 127.0.0.1 as the boulder container cannot connect to the client webserver.  I guess it should be documented that using run-docker.sh or docker-compose.yml require that the client is reachable from a container.
